File: tests/test_spread.py
from fractions import Fraction

from rat_trig.trigonom import spread


def test_spread() -> None:
    """Test spread"""
    # Test with Fraction vectors
    v_1_frac1 = [Fraction(1), Fraction(2)]
    v_2_frac1 = [Fraction(3), Fraction(4)]
    assert spread(v_1_frac1, v_2_frac1) == Fraction(4, 125)

    # Test with float vectors
    v_1_float = [1.0, 2.0]
    v_2_float = [3.0, 4.0]
    assert spread(v_1_float, v_2_float) == 0.032

    # Test with different Fraction vectors
    v_1_frac2 = [Fraction(1, 2), Fraction(1, 4)]
    v_2_frac2 = [Fraction(1, 6), Fraction(1, 8)]
    assert spread(v_1_frac2, v_2_frac2) == Fraction(4, 125)

    # Test with integer vectors (parallel)
    v_1_int1 = [1, 2]
    v_2_int1 = [1, 2]
    assert spread(v_1_int1, v_2_int1) == 0

    # Test with integer vectors (perpendicular)
    v_1_int2 = [1, 0]
    v_2_int2 = [0, 1]
    assert spread(v_1_int2, v_2_int2) == 1


if __name__ == "__main__":
    import pytest

    pytest.main()

File: cpp_ai/tests/test_spread.cpp
/**
 * @file test_spread.cpp
 * @brief Tests for spread function using doctest
 */

#include "doctest.h"
#include "rat_trig/trigonom.hpp"

TEST_CASE("Testing spread function") {
    SUBCASE("Test with integer vectors") {
        std::array<int, 2> vec1 = {1, 2};
        std::array<int, 2> vec2 = {3, 4};
        double expected = 4.0 / 125.0;
        CHECK(rat_trig::spread(vec1, vec2) == doctest::Approx(expected).epsilon(1e-10));
    }

    SUBCASE("Test with float vectors") {
        std::array<float, 2> vec1 = {1.0f, 2.0f};
        std::array<float, 2> vec2 = {3.0f, 4.0f};
        float expected = 4.0f / 125.0f;
        CHECK(rat_trig::spread(vec1, vec2) == doctest::Approx(expected).epsilon(1e-6f));
    }

    SUBCASE("Test with double vectors") {
        std::array<double, 2> vec1 = {1.0, 2.0};
        std::array<double, 2> vec2 = {3.0, 4.0};
        double expected = 4.0 / 125.0;
        CHECK(rat_trig::spread(vec1, vec2) == doctest::Approx(expected).epsilon(1e-10));
    }

    SUBCASE("Test with integer vectors (parallel)") {
        std::array<int, 2> vec1 = {1, 2};
        std::array<int, 2> vec2 = {1, 2};
        CHECK(rat_trig::spread(vec1, vec2) == 0);
    }

    SUBCASE("Test with integer vectors (perpendicular)") {
        std::array<int, 2> vec1 = {1, 0};
        std::array<int, 2> vec2 = {0, 1};
        CHECK(rat_trig::spread(vec1, vec2) == 1);
    }

    SUBCASE("Test with fractions (using doubles)") {
        std::array<double, 2> vec1 = {1.0 / 2.0, 1.0 / 4.0};
        std::array<double, 2> vec2 = {1.0 / 6.0, 1.0 / 8.0};
        double expected = 4.0 / 125.0;
        CHECK(rat_trig::spread(vec1, vec2) == doctest::Approx(expected).epsilon(1e-10));
    }
}
